src/acc/generate_chunk.cpp

```diff
@@ -0,0 +1,167 @@
+
+  // State 1 is always the background state
+  double *energy0 = field.energy0.data;
+  double *density0 = field.density0.data;
+  double *xvel0 = field.xvel0.data;
+  double *yvel0 = field.yvel0.data;
+
+#pragma acc parallel loop gang worker vector default(present) collapse(2) clover_use_target(globals.context.use_target)
+  for (int j = 0; j < (yrange); j++) {
+    for (int i = 0; i < (xrange); i++) {
+      energy0[i + j * base_stride] = state_energy_0;
+      density0[i + j * base_stride] = state_density_0;
+      xvel0[i + j * vels_wk_stride] = state_xvel_0;
+      yvel0[i + j * vels_wk_stride] = state_yvel_0;
+    }
+  }
+
+  for (int state = 1; state < globals.config.number_of_states; ++state) {
+
+    double *cellx = field.cellx.data;
+    double *celly = field.celly.data;
+
+    double *vertexx = field.vertexx.data;
+    double *vertexy = field.vertexy.data;
+
+    const double *state_density = state_density_buffer.data;
+    const double *state_energy = state_energy_buffer.data;
+    const double *state_xvel = state_xvel_buffer.data;
+    const double *state_yvel = state_yvel_buffer.data;
+    const double *state_xmin = state_xmin_buffer.data;
+    const double *state_xmax = state_xmax_buffer.data;
+    const double *state_ymin = state_ymin_buffer.data;
+    const double *state_ymax = state_ymax_buffer.data;
+    const double *state_radius = state_radius_buffer.data;
+    const int *state_geometry = state_geometry_buffer.data;
+
+/* TODO NVHPC says it can't parallelize the inner for loops
+ * We can try forcing if the index calcs are unique */
+#pragma acc parallel loop gang worker vector default(present) collapse(2) clover_use_target(globals.context.use_target)                            \
+    copyin(state_density[ : state_density_buffer.N()]) copyin(state_energy[ : state_energy_buffer.N()])                                \
+    copyin(state_xvel[ : state_xvel_buffer.N()]) copyin(state_yvel[ : state_yvel_buffer.N()])                                          \
+    copyin(state_xmin[ : state_xmin_buffer.N()]) copyin(state_xmax[ : state_xmax_buffer.N()])                                          \
+    copyin(state_ymin[ : state_ymin_buffer.N()]) copyin(state_ymax[ : state_ymax_buffer.N()])                                          \
+    copyin(state_radius[ : state_radius_buffer.N()]) copyin(state_geometry[ : state_geometry_buffer.N()])
+    for (int j = 0; j < (yrange); j++) {
+      for (int i = 0; i < (xrange); i++) {
+        double x_cent = state_xmin[state];
+        double y_cent = state_ymin[state];
+        if (state_geometry[state] == g_rect) {
+          if (vertexx[i + 1] >= state_xmin[state] && vertexx[i] < state_xmax[state]) {
+            if (vertexy[j + 1] >= state_ymin[state] && vertexy[j] < state_ymax[state]) {
+              energy0[i + j * base_stride] = state_energy[state];
+              density0[i + j * base_stride] = state_density[state];
+              for (int kt = j; kt <= j + 1; ++kt) {
```

Comment:
```suggestion
              #pragma acc loop collapse(2)
              for (int kt = j; kt <= j + 1; ++kt) {
```
Since NVHPC could not parallelize the following loops from its own analysis we can "force" the compiler to parallelize the loops by adding a  `#pragma acc loop collapse(2)` to all the nested for loops. 

I passed validation on `clover_bm16_short.in`, but I am not sure if other input decks may result in an error.

Reply:
Be aware that only the following decks have checksums:
```
clover_bm.in
clover_bm16.in
clover_bm16_short.in
clover_bm16_very_short.in
clover_bm2_short.in
clover_bm4_short.in
clover_bm_short.in 
```
In general, if `clover_bm16.in` (this one has 3000 steps so is good at catching memory leaks and MPI related issues) is good at 1 and > 1 MPI ranks with and without staging, then the implementation is correct.